internal/graph/check.go

```diff
@@ -952,6 +960,185 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
+	// TODO: parallelize the lookup.
+	for _, user := range usersetItems {
+		_, visited := info.visitedUserset.LoadOrStore(user, struct{}{})
+		if !visited {
+			newReq := req.clone()
+			newReq.TupleKey.Object = user
+			result, err := recursiveMatchUserUserset(ctx, newReq, info)
+			if err != nil {
+				return nil, err
+			}
+			if result.GetAllowed() {
+				return result, nil
+			}
+		}
+		// Note that visited does not necessary means that there are cycles.  For the following model,
+		// type user
+		// type group
+		//   relations
+		//     define member: [user, group#member]
+		// We have something like
+		// group:1#member@group:2#member
+		// group:1#member@group:3#member
+		// group:2#member@group:a#member
+		// group:3#member@group:a#member
+		// Note that both group:2#member and group:3#member has group:a#member. However, they are not cycles.
+	}
+
+	// There are no intersections and recursive matching shows there are no intersections.
+	return &ResolveCheckResponse{
+		Allowed: false,
+		ResolutionMetadata: &ResolveCheckResponseMetadata{
+			DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + info.dsCount.Load(),
+		},
+	}, nil
+}
+
+func nestedUsersetFastpath(ctx context.Context,
+	typesys *typesystem.TypeSystem,
+	ds storage.RelationshipTupleReader,
+	req *ResolveCheckRequest) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "nestedUsersetFastpath")
+	defer span.End()
+
+	// Note that this will take care of the publicly assignable case as well.
+	iter, err := checkutil.IteratorReadStartingFromUser(ctx,
+		typesys, ds, req,
+		tuple.ToObjectRelationString(tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()),
+		nil)
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+	filteredIter := storage.NewConditionsFilteredTupleKeyIterator(
+		storage.NewFilteredTupleKeyIterator(
+			storage.NewTupleKeyIteratorFromTupleIterator(iter),
+			validation.FilterInvalidTuples(typesys),
+		),
+		checkutil.BuildTupleKeyConditionFilter(ctx, req.GetContext(), typesys),
+	)
+	defer filteredIter.Stop()
+
+	group, err := storage.GetAllObjects(ctx, filteredIter)
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+
+	if group.Size() == 0 {
+		return &ResolveCheckResponse{
+			Allowed: false,
+			ResolutionMetadata: &ResolveCheckResponseMetadata{
+				DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 1,
+			},
+		}, nil
+	}
+
+	if group.Exists(req.GetTupleKey().GetObject()) {
+		return &ResolveCheckResponse{
+			Allowed: true,
+			ResolutionMetadata: &ResolveCheckResponseMetadata{
+				DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 1,
+			},
+		}, nil
+	}
+
+	reqTupleKey := req.GetTupleKey()
+	objectType := tuple.GetType(reqTupleKey.GetObject())
+	relation := reqTupleKey.GetRelation()
+
+	dsCount := &atomic.Uint32{}
+	dsCount.Store(1)
+	return recursiveMatchUserUserset(ctx, req, &recursiveMatchUserUsersetInfo{
```

Comment:
We could start processing this without having to wait for all the Objects to be loaded into memory?